wlroots/wlr_types/layer_shell_v1.py

```diff
@@ -0,0 +1,213 @@
+from weakref import WeakKeyDictionary
+
+from pywayland.server import Signal
+
+from wlroots import ffi, lib, Ptr
+from .output import Output
+from .surface import Surface
+from .xdg_shell import SurfaceCallback, T
+
+if TYPE_CHECKING:
+    from pywayland.server import Display
+
+_weakkeydict: WeakKeyDictionary = WeakKeyDictionary()
+
+
+class LayerSurfaceV1KeyboardInteractivity(enum.IntEnum):
+    NONE = 0
+    EXCLUSIVE = 1
+    ON_DEMAND = 2
+
+
+class LayerSurfaceV1Anchor(enum.IntFlag):
+    TOP = 1
+    BOTTOM = 2
+    LEFT = 4
+    RIGHT = 8
+    VERTICAL = TOP | BOTTOM
+    HORIZONTAL = LEFT | RIGHT
+
+
+class LayerShellV1Layer(enum.IntEnum):
+    BACKGROUND = 0
+    BOTTOM = 1
+    TOP = 2
+    OVERLAY = 3
+
+
+class Margin:
+    def __init__(self, top, right, bottom, left):
+        self.top = top
+        self.right = right
+        self.bottom = bottom
+        self.left = left
+
+
+class LayerSurfaceV1State(Ptr):
+    def __init__(self, ptr):
+        self._ptr = ptr
+
+    @property
+    def anchor(self) -> LayerSurfaceV1Anchor:
+        return LayerSurfaceV1Anchor(self._ptr.anchor)
+
+    @property
+    def exclusive_zone(self) -> int:
+        return self._ptr.exclusive_zone
+
+    @property
+    def margin(self) -> Margin:
+        margin = self._ptr.margin
+        return Margin(margin.top, margin.right, margin.bottom, margin.left)
+
+    @property
+    def keyboard_interactive(self) -> LayerSurfaceV1KeyboardInteractivity:
+        return LayerSurfaceV1KeyboardInteractivity(self._ptr.keyboard_interactive)
+
+    @property
+    def desired_width(self) -> int:
+        return self._ptr.desired_width
+
+    @property
+    def desired_height(self) -> int:
+        return self._ptr.desired_height
+
+    @property
+    def actual_width(self) -> int:
+        return self._ptr.actual_width
+
+    @property
+    def actual_height(self) -> int:
+        return self._ptr.actual_height
+
+    @property
+    def layer(self) -> LayerShellV1Layer:
+        return LayerShellV1Layer(self._ptr.layer)
+
+
+class LayerSurfaceV1(Ptr):
+    def __init__(self, ptr):
+        self._ptr = ffi.cast("struct wlr_layer_surface_v1 *", ptr)
+
+        self.destroy_event = Signal(ptr=ffi.addressof(self._ptr.events.destroy))
+        self.map_event = Signal(ptr=ffi.addressof(self._ptr.events.map))
+        self.unmap_event = Signal(ptr=ffi.addressof(self._ptr.events.unmap))
+        self.new_popup_event = Signal(ptr=ffi.addressof(self._ptr.events.new_popup))
+
+    @property
+    def surface(self) -> Surface:
+        surface_ptr = self._ptr.surface
+        _weakkeydict[surface_ptr] = self._ptr
+        return Surface(surface_ptr)
+
+    @property
+    def output(self) -> Optional[Output]:
+        output_ptr = self._ptr.output
+        if output_ptr == ffi.NULL:
+            return None
+        _weakkeydict[output_ptr] = self._ptr
+        return Output(output_ptr)
+
+    @output.setter
+    def output(self, output: Output):
```

Comment:
Would you ever want to remove the output by setting it to `None`?

Reply:
Hmm, I don't think so. Layout surfaces are bound to an output, and you would destroy them if you were destroying the output.